### PR TITLE
Add generator function for the js client + associated test

### PR DESCRIPTION
**Pull request by GaetanMil**
close : https://github.com/craft-ai/craft-ai-client-js/issues/177

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -36,6 +36,8 @@ const isUnvalidConfiguration = (configuration) =>
   _.isUndefined(configuration) || !_.isObject(configuration);
 const areUnvalidOperations = (operations) =>
   _.isUndefined(operations) || !_.isArray(operations);
+const isInvalidFilter = (filter) => 
+  _.isArray(filter) && filter.reduce((allValid, agentName) => !AGENT_ID_ALLOWED_REGEXP.test(agentName) && allValid, true);
```
Comment (Totoketchup) [inline src/client.js]: You could use Array.every() here I guess

File: src/client.js [hunk 2]
```diff
@@ -628,8 +630,169 @@ export default function createClient(tokenOrCfg) {
           decision.timestamp = posixTimestamp;
           return decision;
         });
+    },
+    // Generators methods
+    createGenerator: function(configuration, filter, generatorName) {
+      if (isUnvalidConfiguration(configuration)) {
+        return Promise.reject(
+          new CraftAiBadRequestError(
+            'Bad Request, unable to create an generator with no or invalid configuration provided.'
```
Comment (Totoketchup) [inline src/client.js]:
```suggestion
            'Bad Request, unable to create a generator with no or invalid configuration provided.'
```

File: src/client.js [hunk 3]
```diff
@@ -628,8 +630,169 @@ export default function createClient(tokenOrCfg) {
...
+          )
+        );
+      }
+
+      if (isUnvalidId(generatorName)) {
+        return Promise.reject(
+          new CraftAiBadRequestError(
+            `Bad Request, unable to create an generator with invalid agent id. It must only contain characters in 'a-zA-Z0-9_-' and must be a string between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
```
Comment (Totoketchup) [inline src/client.js]:
```suggestion
            `Bad Request, unable to create a generator with an invalid agent id. It must only contain characters in 'a-zA-Z0-9_-' and must be a string between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
```

File: src/client.js [hunk 4]
```diff
@@ -628,8 +630,169 @@ export default function createClient(tokenOrCfg) {
...
+          )
+        );
+      }
+
+      if (isInvalidFilter(filter)) {
+        return Promise.reject(
+          new CraftAiBadRequestError(
+            `Bad Request, unable to create an generator with invalid filter. It must be a list of string containing characters in 'a-zA-Z0-9_-' and must be between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
```
Comment (Totoketchup) [inline src/client.js]:
```suggestion
            `Bad Request, unable to create a generator with an invalid filter. It must be a list of string containing characters in 'a-zA-Z0-9_-' and must be between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
```

File: src/client.js [hunk 5]
```diff
@@ -628,8 +630,169 @@ export default function createClient(tokenOrCfg) {
...
+        .then(({ body }) => {
+          debug(`Generator '${body.id}' created.`);
+          return body;
+        });
+    },
+    deleteGenerator: function(generatorName) {
+      if (isUnvalidId(generatorName)) {
+        return Promise.reject(
+          new CraftAiBadRequestError(
+            `Bad Request, unable to delete an generator with invalid generator id. It must only contain characters in 'a-zA-Z0-9_-' and must be a string between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
```
Comment (Totoketchup) [inline src/client.js]:
```suggestion
            `Bad Request, unable to delete a generator with an invalid generator id. It must only contain characters in 'a-zA-Z0-9_-' and must be a string between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
```

File: src/client.js [hunk 6]
```diff
@@ -628,8 +630,169 @@ export default function createClient(tokenOrCfg) {
...
+        path: `/generators/${generatorName}`
+      })
+        .then(({ body }) => {
+          debug(`Generator '${generatorName}' deleted`);
+          return body;
+        });
+    },
+    getGeneratorTree: function(
+      generatorName,
+      t = undefined,
```
Comment (Totoketchup) [inline src/client.js]: Default would be undefined, then there is no need to put here.

File: src/client.js [hunk 7]
```diff
@@ -628,8 +630,169 @@ export default function createClient(tokenOrCfg) {
...
+      version = DEFAULT_DECISION_TREE_VERSION
+    ) {
+      if (isUnvalidId(generatorName)) {
+        return Promise.reject(
+          new CraftAiBadRequestError(
+            `Bad Request, unable to get a decision tree with invalid generator id. It must only contain characters in 'a-zA-Z0-9_-' and must be a string between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
```
Comment (Totoketchup) [inline src/client.js]:
```suggestion
            `Bad Request, unable to get a decision tree from an invalid generator id. It must only contain characters in 'a-zA-Z0-9_-' and must be a string between 1 and ${AGENT_ID_MAX_LENGTH} characters.`
```

File: src/client.js [hunk 8]
```diff
@@ -628,8 +630,169 @@ export default function createClient(tokenOrCfg) {
...
+          )
+        );
+      }
+      let posixTimestamp = Time(t).timestamp;
+      if (_.isUndefined(posixTimestamp)) {
+        return Promise.reject(
+          new CraftAiBadRequestError(
+            'Bad Request, unable to retrieve an agent decision tree with an invalid timestamp provided.'
```
Comment (Totoketchup) [inline src/client.js]: Generator here no?

File: src/client.js (hunk 1 above)
Comment (GaetanMil) [reply src/client.js]: Yes 👍

File: src/client.js (hunk 6 above)
Comment (GaetanMil) [reply src/client.js]: It's what is done in almost every function that has optional parameters 
See per example : 
- `getAgentInspectorUrl: function(agentId, t = undefined)`
- `getAgentDecisionTree: function(agentId, t = undefined, version=DEFAULT_DECISION_TREE_VERSION)`
- ...
Maybe I can update it everywhere

File: test/data/configuration_1_generator.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    }
+  },
+  "output": [
+    "lightbulbColor"
+  ],
+  "time_quantum": 100,
+  "learning_period": 1500000,
+  "operations_as_events": true,
+  "tree_max_operations": 10000
+}
```
Comment (Totoketchup) [inline test/data/configuration_1_generator.json]: add new line

File: test/test_getGeneratorDecision.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      lightIntensity: 1
+    }, {
+      lightIntensity: 0.1
+    })
+      .then((decision) => {
+        expect(decision).to.be.ok;
+        expect(decision.output.lightbulbColor.predicted_value).to.be.equal('black');
+      });
+  });
+});
```
Comment (Totoketchup) [inline test/test_getGeneratorDecision.js]: newline

File: test/test_createGenerator.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+  let client;
+
+  const VALID_FILTER = ['agent_1', 'agent_2'];
+  const VALID_GENERATOR_NAME = 'generator_1';
+  before(function() {
+    client = craftai(CRAFT_CFG);
+    expect(client).to.be.ok;
+  });
+
+  it('Should works with a configuration and valid generator name provided', () => {
```
Comment (Totoketchup) [inline test/test_createGenerator.js]:
```suggestion
  it('Should work with a configuration and valid generator name provided', () => {
```

File: test/test_createGenerator.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      })
+      .catch((err) => {
+        client.deleteAgent(VALID_GENERATOR_NAME) // The test might fail due to duplicate id, let's make sure it doesn't fail twice.
+          .then(() => {
+            throw err;
+          });
+      });
+  });
+
+  it('Should not work when asked twice', () => {
```
Comment (Totoketchup) [inline test/test_createGenerator.js]:
```suggestion
  it('Should not work when creating the same generator is asked twice', () => {
```

File: test/test_createGenerator.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+      .then((res) => 
+        Promise.reject(new Error(`Should not be reached but as the result : ${res}`))
+      )
+      .catch((err) => {
+        expect(err).to.be.an.instanceof(errors.CraftAiError);
+        expect(err).to.be.an.instanceof(errors.CraftAiBadRequestError);
+      });
+  });
+
+  it('Should fail with an undefined configuration', () => {
```
Comment (Totoketchup) [inline test/test_createGenerator.js]:
```suggestion
  it('Should fail when created with an undefined configuration', () => {
```

File: test/test_createGenerator.js [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+      .then((res) => 
+        Promise.reject(new Error(`Should not be reached but as the result : ${res}`))
+      )
+      .catch((err) => {
+        expect(err).to.be.an.instanceof(errors.CraftAiError);
+        expect(err).to.be.an.instanceof(errors.CraftAiBadRequestError);
+      });
+  });
+
+  it('Should fail with an invalid configuration', () => {
```
Comment (Totoketchup) [inline test/test_createGenerator.js]:
```suggestion
  it('Should fail when created with an invalid configuration', () => {
```

File: test/test_createGenerator.js (hunk 11 above)
Comment (Totoketchup) [reply test/test_createGenerator.js]: I don't remember how others `it` are done, but does it begin with capital letters ? If no, then whenge this back to small letters ;)

File: test/test_getGeneratorDecision.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+  it('should succeed when using valid parameters', function() {
+    return client.computeGeneratorDecision(GENERATOR_NAME, CONFIGURATION_1_OPERATIONS_1_TO, {
+      presence: 'none',
+      lightIntensity: 0.1
+    })
+      .then((decision) => {
+        expect(decision).to.be.ok;
+        expect(decision.output.lightbulbColor.predicted_value).to.be.equal('black');
+      });
+  });
```
Comment (Totoketchup) [inline test/test_getGeneratorDecision.js]:
```suggestion
  });
  
```

File: test/test_getGeneratorDecision.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        return client.addAgentContextOperations(AGENT_NAME, CONFIGURATION_1_OPERATIONS_1);
+      })
+      .then(() => client.createGenerator(CONFIGURATION_1_GENERATOR, VALID_FILTER, GENERATOR_NAME))
+      .catch((err) => console.log('What the fuck :', err));
+  });
+
+  afterEach(function() {
+    return client.deleteAgent(AGENT_NAME);
+  });
+
```
Comment (Totoketchup) [inline test/test_getGeneratorDecision.js]: Can you test that it fails on some conditions too ?

File: test/test_getGeneratorDecisionTree.js [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+describe('client.getGeneratorDecisionTree(<agentId>, <timestamp>)', function() {
+  let client;
+  let agent;
+  const AGENT_NAME = `get_generator_decision_${RUN_ID}`;
+  const GENERATOR_NAME = `generator_${RUN_ID}`;
+  const CONFIGURATION_1_OPERATIONS_1_TO = _.last(CONFIGURATION_1_OPERATIONS_1).timestamp;
+  const VALID_FILTER = [AGENT_NAME];
+  let CONFIGURATION_1_GENRATOR_IN_DB = _.cloneDeep(CONFIGURATION_1_GENERATOR);
+  CONFIGURATION_1_GENRATOR_IN_DB.filter = VALID_FILTER;
```
Comment (Totoketchup) [inline test/test_getGeneratorDecisionTree.js]:
```suggestion
  CONFIGURATION_1_GENRATOR_IN_DB.filter = VALID_FILTER;
  
```

File: test/test_getGeneratorDecisionTree.js [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+  before(function() {
+    client = craftai(CRAFT_CFG);
+    expect(client).to.be.ok;
+    return client.deleteAgent(AGENT_NAME)
+      .then((res) => {
```
Comment (Totoketchup) [inline test/test_getGeneratorDecisionTree.js]: no need of return and brackets

File: test/test_getGeneratorDecisionTree.js [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+import craftai, { errors } from '../src';
+
+describe('client.getGeneratorDecisionTree(<agentId>, <timestamp>)', function() {
+  let client;
+  let agent;
+  const AGENT_NAME = `get_generator_decision_${RUN_ID}`;
+  const GENERATOR_NAME = `generator_${RUN_ID}`;
+  const CONFIGURATION_1_OPERATIONS_1_TO = _.last(CONFIGURATION_1_OPERATIONS_1).timestamp;
+  const VALID_FILTER = [AGENT_NAME];
+  let CONFIGURATION_1_GENRATOR_IN_DB = _.cloneDeep(CONFIGURATION_1_GENERATOR);
```
Comment (Totoketchup) [inline test/test_getGeneratorDecisionTree.js]:
```suggestion
  let CONFIGURATION_1_GENERATOR_IN_DB = _.cloneDeep(CONFIGURATION_1_GENERATOR);
```

File: test/test_getGeneratorDecisionTree.js [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+  afterEach(() => client.deleteGenerator(GENERATOR_NAME));
+
+  it('Success on valid parameters', () => {
+    return client.getGeneratorTree(GENERATOR_NAME, CONFIGURATION_1_OPERATIONS_1_TO)
+      .then((treeJson) => {
+        expect(treeJson).to.be.ok;
+        const { _version, configuration, trees } = parse(treeJson);
+        expect(trees).to.be.ok;
+        expect(_version).to.be.ok;
+        expect(configuration).to.be.deep.equal(CONFIGURATION_1_GENRATOR_IN_DB);
```
Comment (Totoketchup) [inline test/test_getGeneratorDecisionTree.js]:
```suggestion
        expect(configuration).to.be.deep.equal(CONFIGURATION_1_GENERATOR_IN_DB);
```

File: test/test_listGenerators.js [hunk 21]
```diff
@@ -0,0 +1,37 @@
+import CONFIGURATION_AGENT from './data/configuration_1.json';
+import CONFIGURATION_GENERATOR from './data/configuration_1_generator.json';
+
+import craftai from '../src';
+
+describe('client.listGenerators()', function() {
+  let client;
+  const generatorsId = [`list_generators_${RUN_ID}_1`, `list_generators_${RUN_ID}_2`, `list_generators_${RUN_ID}_3`];
```
Comment (Totoketchup) [inline test/test_listGenerators.js]:
```suggestion
  const GENERATORS_ID = [`list_generators_${RUN_ID}_1`, `list_generators_${RUN_ID}_2`, `list_generators_${RUN_ID}_3`];
```

File: test/test_listGenerators.js [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+  const agentId = `testAgentListGenerators${RUN_ID}`;
```
Comment (Totoketchup) [inline test/test_listGenerators.js]:
```suggestion
  const AGENT_ID = `testAgentListGenerators${RUN_ID}`;
```

File: test/test_listGenerators.js [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+  before(function() {
+    client = craftai(CRAFT_CFG);
+    expect(client).to.be.ok;
+    return client.createAgent(CONFIGURATION_AGENT, agentId);
+  });
+
+  beforeEach(function() {
+    return Promise.all(_.map(generatorsId, (generatorId) => client.deleteGenerator(generatorId))) // Delete any preexisting generator with this id.
+      .then(() => Promise.all(_.map(generatorsId, (generatorId) => client.createGenerator(CONFIGURATION_GENERATOR,[agentId], generatorId))));
```
Comment (Totoketchup) [inline test/test_listGenerators.js]:
```suggestion
      .then(() => Promise.all(_.map(generatorsId, (generatorId) => client.createGenerator(CONFIGURATION_GENERATOR, [agentId], generatorId))));
```

File: test/test_createGenerator.js (hunk 12 above)
Comment (GaetanMil) [reply test/test_createGenerator.js]: There is already createGenerator in the name of the describe

**Pull request merged**